api/server/middleware_auth.go

```diff
@@ -313,24 +314,43 @@ func (a *authMiddleware) parseSecret(
 ) (string, string, error) {
 	if a.provider.Type() == secrets.TypeK8s && fetchCOLabels {
 		// For k8s fetch the actual annotations
-		pvcName, ok := locatorLabels[PVCNameLabelKey]
-		if !ok {
-			// best effort to fetch the secret
-			return parseSecretFromLabels(specLabels, locatorLabels)
-		}
-		pvcNamespace, ok := locatorLabels[PVCNamespaceLabelKey]
-		if !ok {
+		pvcName, nameFound := locatorLabels[PVCNameLabelKey]
+		pvcNamespace, namespaceFound := locatorLabels[PVCNamespaceLabelKey]
+		if !nameFound || !namespaceFound {
 			// best effort to fetch the secret
 			return parseSecretFromLabels(specLabels, locatorLabels)
 		}
 
+		// Get PVC for checking secrets in the annotations
 		pvc, err := k8s.Instance().GetPersistentVolumeClaim(pvcName, pvcNamespace)
 		if err != nil {
 			return "", "", err
 		}
-		secretName := pvc.ObjectMeta.Annotations[secrets.SecretNameKey]
-		secretNamespace := pvc.ObjectMeta.Annotations[secrets.SecretNamespaceKey]
-		return secretName, secretNamespace, nil
+
+		// Get SC for checking secrets in annotations
+		scName := pvc.ObjectMeta.Annotations[v1.BetaStorageClassAnnotation]
+		sc, err := k8s.Instance().GetStorageClass(scName)
+		if err != nil {
+			return "", "", err
+		}
+
+		// Get secret name from PVC first. If does not exists, get from SC.
+		secretName, nameFound := pvc.ObjectMeta.Annotations[secrets.SecretNameKey]
+		if !nameFound {
+			secretName, nameFound = sc.ObjectMeta.Annotations[secrets.SecretNameKey]
```

Comment:
I thought the secret name will be provided as a storage class parameter and not as storage class annotations. @lpabon Do we want to read annotations from storage class like this?

Reply:
As discussed offline, this info will be translated from storage class params to spec.VolumeLabels, so this PR is not needed.